integration-tests/conditional-compilation/lib.rs

```diff
@@ -0,0 +1,92 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::new_without_default)]
+
+#[ink::trait_definition]
+pub trait Flip {
+    /// Flips the current value of the Flipper's boolean.
+    #[ink(message)]
+    fn flip(&mut self);
+
+    /// Returns the current value of the Flipper's boolean.
+    #[ink(message)]
+    fn get(&self) -> bool;
+
+    #[cfg(feature = "foo")]
+    #[ink(message)]
+    fn push_foo(&mut self, value: bool);
+}
+
+#[ink::contract]
+pub mod conditional_compilation {
+    use super::Flip;
+
+    /// Feature gated event
+    #[cfg(feature = "foo")]
+    #[ink(event)]
+    pub struct Changes {
+        // attributing event field with `cfg` is not allowed
+        new_value: bool,
+        #[ink(topic)]
+        by: AccountId,
+    }
+
+    #[ink(storage)]
+    pub struct ConditionalCompilation {
+        value: bool,
+    }
+
+    impl ConditionalCompilation {
+        /// Creates a new flipper smart contract initialized to `false`.
+        #[ink(constructor)]
+        pub fn new() -> Self {
```

Comment:
Also, it would be nice to have several constructors and messages here compiled conditionally=)